parmed/formats/sdf.py

```diff
@@ -0,0 +1,44 @@
+"""
+This module contains parsers for sdf file format 
+extension described at https://www.cas.org/content/chemical-suppliers/example-sdf
+"""
+from __future__ import print_function, division, absolute_import
+
+from parmed.formats.registry import FileFormatType
+from parmed.utils.io import genopen
+from parmed.utils.six import add_metaclass
+from parmed import rdkit
+
+@add_metaclass(FileFormatType)
+class SDFFile(object):
+    """ Class to read SDF file """
+
+    @staticmethod
+    def id_format(filename):
+        """ Identify the file as a SDF file format or not
+
+        Parameters
+        ----------
+        filename : str
+            Name of the file to test whether or not it is a sdf file
+
+        Returns
+        -------
+        is_fmt : bool
+            True if it is a sdf file, False otherwise
+        """
+        f = genopen(filename, 'r')
+        try:
+            for line in f:
+                if not line.strip(): continue
+                if line.startswith('M  END') or line.startswith('$$$$'):
+                    return True
+                else:
+                    continue
```

Comment:
The `FileFormatType` metaclass checks the attributes of a class after it's been constructed and checks to see if there is a `id_format` method implemented on that class.  If there is, it adds the newly constructed class to a parser registry.

The `parmed.load_file` then iterates through the registry and passes the file to the `id_format` method until one of the classes returns `True` (indicating that the file format has been identified).  If that class has a `parse` method, it then passes the filename to that `parse` method.  Otherwise, it passes the filename to the main constructor.

Since the registry is a dict (and you can never be sure what order the classes are constructed in the first place), there is no well-defined order for which class formats are identified first.  So you have to assume that any class's `id_format` method will be called for any kind of file.  That's why it's important that `id_format` be able to tell, as best as possible, whether a file is a particular format or not within <20 lines or so.  Most classes do it within 5 lines.

Also note that the format ID doesn't have to be perfect -- it just has to catch *most* of the use-cases (for some definition of "most").  I would base the format ID on the header of each section and stipulate that it must occur within 5 lines or so (and only ignore blank lines or ones with valid comments, if that exists).